cloudify_cli/commands/apply.py

```diff
@@ -0,0 +1,188 @@
+from ..blueprint import get_blueprint_path_and_id
+from . import blueprints, install, deployments
+
+
+@cfy.command(name='apply',
+             short_help='Install a blueprint or update existing deployment '
+                        'with blueprint [manager only]')
+@cfy.argument('blueprint-path')
+@cfy.argument('deployment-id')
+@cfy.options.blueprint_filename()
+@cfy.options.blueprint_id()
+@cfy.options.inputs
+@cfy.options.reinstall_list
+@cfy.options.workflow_id()
+@cfy.options.skip_install
+@cfy.options.skip_uninstall
+@cfy.options.dont_skip_reinstall
+@cfy.options.ignore_failure
+@cfy.options.install_first
+@cfy.options.preview
+@cfy.options.dont_update_plugins
+@cfy.options.force(help=helptexts.FORCE_UPDATE)
+@cfy.options.tenant_name(required=False,
+                         resource_name_for_help='blueprint and deployment')
+@cfy.options.visibility(mutually_exclusive_required=False)
+@cfy.options.validate
+@cfy.options.include_logs
+@cfy.options.json_output
+@cfy.options.common_options
+@cfy.options.runtime_only_evaluation
+@cfy.options.auto_correct_types
+@cfy.options.reevaluate_active_statuses()
+# From install command
+@cfy.options.skip_plugins_validation
+@cfy.options.parameters
+@cfy.options.allow_custom_parameters
+@cfy.assert_manager_active()
+@cfy.pass_client()
+@cfy.pass_logger
+@cfy.pass_context
+def apply(ctx,
+          blueprint_path,
+          deployment_id,
+          blueprint_filename,
+          blueprint_id,
+          inputs,
+          reinstall_list,
+          workflow_id,
+          skip_install,
+          skip_uninstall,
+          dont_skip_reinstall,
+          ignore_failure,
+          install_first,
+          preview,
+          dont_update_plugins,
+          force,
+          tenant_name,
+          visibility,
+          validate,
+          include_logs,
+          json_output,
+          runtime_only_evaluation,
+          auto_correct_types,
+          reevaluate_active_statuses,
+          skip_plugins_validation,
+          parameters,
+          allow_custom_parameters,
+          logger,
+          client
+          ):
+    """Apply command uses cfy install or deployment update depends on
+    existence of DEPLOYMENT_ID deployment.
+
+    If the deployment exists, the deployment will be updated with the
+    given blueprint.
+    otherwise the blueprint will installed (the deployment name will be
+    DEPLOYMENT_ID).
+    In both cases the blueprint is being uploaded to the manager.
+
+    `BLUEPRINT_PATH` can be a:
+        - local blueprint yaml file
+        - blueprint archive
+        - url to a blueprint archive
+        - github repo (`organization/blueprint_repo[:tag/branch]`)
+
+    Supported archive types are: zip, tar, tar.gz and tar.bz2
+
+    `DEPLOYMENT_ID` is the deployment's id to install/update.
+    """
+    # check if deployment exists
+    if deployment_id not in [deployment.id for deployment in
+                             client.deployments.list()]:
```

Comment:
this won't work because
1. `client.deployments.list()` returns deployment objects and not just ids
2. it is paginated so it will return False even though the deployment might exist, just not be in the first 1000 returned

use client.deployments.get with try/except 404 instead

Reply:
oh i see that it does `deployment.id` so scratch 1, but 2 still applies

Reply:
Done.